libp2p/stream/chronosstream.nim

```diff
@@ -0,0 +1,109 @@
+## Nim-LibP2P
+## Copyright (c) 2018 Status Research & Development GmbH
+## Licensed under either of
+##  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+##  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+## at your option.
+## This file may not be copied, modified, or distributed except according to
+## those terms.
+
+import chronos, chronicles
+import lpstream
+
+logScope:
+  topic = "ChronosStream"
+
+type ChronosStream* = ref object of LPStream
+    reader: AsyncStreamReader
+    writer: AsyncStreamWriter
+    server: StreamServer
+    client: StreamTransport
+
+proc newChronosStream*(server: StreamServer, 
+                       client: StreamTransport): ChronosStream =
+  new result
+  result.server = server
+  result.client = client
+  result.reader = newAsyncStreamReader(client)
+  result.writer = newAsyncStreamWriter(client)
+  result.closed = false
+
+method read*(s: ChronosStream, n = -1): Future[seq[byte]] {.async, gcsafe.} = 
+  try:
+    result = await s.reader.read(n)
+  except AsyncStreamReadError as exc:
+    raise newLPStreamReadError(exc.par)
+
+method readExactly*(s: ChronosStream, 
+                    pbytes: pointer, 
+                    nbytes: int): Future[void] {.async, gcsafe.} =
+  try:
+    await s.reader.readExactly(pbytes, nbytes)
+  except AsyncStreamIncompleteError:
+    raise newLPStreamIncompleteError()
+  except AsyncStreamReadError as exc:
+    raise newLPStreamReadError(exc.par)
+
+method readLine*(s: ChronosStream, limit = 0, sep = "\r\n"): Future[string] {.async, gcsafe.} =
+    try:
+      result = await s.reader.readLine(limit, sep)
+    except AsyncStreamReadError as exc:
+      raise newLPStreamReadError(exc.par)
+
+method readOnce*(s: ChronosStream, pbytes: pointer, nbytes: int): Future[int] {.async, gcsafe.} =
+    try:
+      result = await s.reader.readOnce(pbytes, nbytes)
+    except AsyncStreamReadError as exc:
+      raise newLPStreamReadError(exc.par)
+
+method readUntil*(s: ChronosStream, 
+                  pbytes: pointer, 
+                  nbytes: int, 
+                  sep: seq[byte]): Future[int] {.async, gcsafe.} =
+  try:
+    result = await s.reader.readUntil(pbytes, nbytes, sep)
+  except TransportIncompleteError, AsyncStreamIncompleteError:
```

Comment:
I should not be intercepting transport exceptions here, only stream exceptions - I'll remove it.

Reply:
why is `asyncstreamincompleteerror` (when a file or buffer stream reaches EOF, from what I can tell) different from when a transport in a socket stream is closed (the equivalent of EOF, really)?

Reply:
Yep... This is a problem in general. I'm not happy about the transport vs stream separation. I either deal with a stream or a raw socket, but both transport and stream have a lot of competing abstractions. I want to rethink that down the line.